### DISCUSSION TITLE
How to create a scalar index on a boolean field

### DISCUSSION THREAD

**Discussion by ducanh997** [milvus-io/milvus#31260] [Q&A and General discussion] [answered]
I have a collection with a boolean field, and I occasionally need to filter records based on this field. My question is how to create an index for this specific field?. When attempting to index the data using the below code snippet, I encountered an exception:

`create index on Bool field is not supported: invalid parameter[expected=supported field][actual=create index on Bool field]`

Additionally, I'm wondering about the comparative advantages between utilizing an index and employing a partition column. Any insights into the preferred approach would be greatly appreciated.

Thank you in advance for your assistance.

**Comment by yhmo** [comment, marked as answer]
Scalar index only supports VARVAR, integer, and float type fields. Bool type is not supported.
The doc: https://milvus.io/docs/scalar_index.md#Scalar-Index is wrong, we will fix it.


Milvus manages data by segment. Query node searches segments one by one. In each segment, a vector field has an independent index, a scalar field also has an independent index if you create it. Index is to accelerate search speed within a segment. When you call collection.search() to do ANN search, the vector field index is used. If you add a filter expression for collection.search(expr="a > 10"), the scalar index of the "a" field is used.

Split data into partitions, then you control the search scope in a few partitions to improve search performance. 
A collection contains multiple partitions. A partition contains multiple segments.

**Reply by ducanh997** [reply]
Hi @yhmo.
Could you explain more on how Milvus do a hybrid search internally? How does Milvus combine two types of indexes to search within a segment.

**Reply by yhmo** [reply]
Search with filter expression, the baisc logic:
1. build a bitset by the filter expression: https://milvus.io/docs/bitset.md#Overview
Assume we have 3 rows in a segment:
{"id": 1, "vector": [...], "age": 23},
{"id": 2, "vector": [...], "age": 56},
{"id": 3, "vector": [...], "age": 12},

The "age" field is integer type. If you create an index for the "age" field, the index is a map whose key is simply sorted: {12:2, 23:0, 56:1}.
If you search with an expression "age == 12", the parser quickly knows the 12 is row_2.
Then it creates a bitset list: [0, 0, 1], which means the row_0 and row_1 are not matched this expression, the row_2 matches the expression.
If you search with "age > 30", then then items whose key value is less than 30 will be marked as "0". The bitset list is [0, 1, 0]

2. pass the bitset to the search engine, the items whose flag is 0 will be ignored, the search engine only computes vector distance for the items whose flag is 1. No matter IVF vector index or HNSW index, the items whose flag is 0 will be ignored.

**Reply by ducanh997** [reply]
Thanks for providing such a clear explanation

**Reply by ducanh997** [reply]
Hi @yhmo. I have another question. 

My collection has a status field with the type of boolean, so I have three options:
1. Convert this field to int and create an index to improve hybrid search performance.
2. Explicitly create two partition for the collection.
3. Create a status partition field.

Which method is more effective for handling an imbalanced dataset containing 50 million records, where only 1% of them have a False value in the status field?

**Reply by yhmo** [reply]
If the status has only True or False, "two partitions" approach is better.
